### PR TITLE
utils: implement helper for Transformers weight loader

### PR DESCRIPTION
**Pull request by fa-ina-tic**
This PR adds helper function to load weight files from model directory and return its path.
This enables loading weight files from local machine.

p.s. Since ``snapshot_download`` is not yet implemented, loading weights from Hugging Face Hub is not yet supported.

### CODE REVIEW

File: mako/utils/transformers.cc [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#include <iostream>
+#include <fstream>
+#include <tuple>
+#include <string>
+#include <filesystem>
+#include <algorithm>
+
+namespace fs = std::filesystem;
+
+/// \brief Utility to find weight files from model directory
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please finish commenting for ``prepare_hf_model_weights``.

File: mako/utils/transformers.cc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+/// \param model_name_or_path A path to a directory containing model weights saved using
+/// \param
+/// \param
+/// \param
+/// \param
+/// \return
+std::tuple<std::string, std::vector<std::string>, bool> prepare_hf_model_weights (
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please use ``static inline`` primitives before static functions. Also, remove black space between function name and bracket.
i.e., ``static inline std::tuple<std::string, std::vector<std::string>, bool> prepare_hf_model_weights(``

File: mako/utils/transformers.cc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        std::string model_name_or_path,
+        std::string cache_dir = ,
```
Comment (9rum) [inline mako/utils/transformers.cc]: Syntax error

File: mako/utils/transformers.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        std::string load_format = "auto",
+        bool fall_back_to_pt = true,
+        std::string revision = "None") {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please use ``std::optional`` for type of a nullable parameter and ``std::nullopt`` for its default value.

File: mako/utils/transformers.cc [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    std::vector<std::string> allow_patterns;
+    try {
+        std::vector<std::string> format_allowance = {"auto", "safetensors", "pt", "npcache"};
+        if (std::find(format_allowance.begin(), format_allowance.end(), load_format) == format_allowance.end())
+            throw std::invalid_argument("invalid load_format, it should be one of : auto, safetensors, pt, npcache");
+    } catch(std::invalid_argument &e) {
+        std::cerr << e.what();
+    }
+
+    if (load_format == "auto") {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please use ``std::string::compare`` for string comparison.

File: mako/utils/transformers.cc [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        allow_patterns = {".bin"};
+    }
+
+    if (fall_back_to_pt) {
+        allow_patterns.push_back(".pt");
+    }
+
+    std::string hf_folder;
+    if (!is_local) {
+        // TODO(soomin): implemen snapshot_download
```
Comment (9rum) [inline mako/utils/transformers.cc]: Fix typo: implemen => implement

File: mako/utils/transformers.cc [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        if (!hf_weights_files.empty()) {
+            if (pattern == ".safetensors") {
+                use_safetensors = true;
+            }
+            break;
+        }
+    }
+    // exclude file
```
Comment (9rum) [inline mako/utils/transformers.cc]: I think this comment is ambiguous.

File: mako/utils/transformers.cc [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    if (!is_local) {
+        // TODO(soomin): implemen snapshot_download
+    } else {
+        hf_folder = model_name_or_path;
+    }
+
+    // load weight files path from model path directory
+    std::vector<std::string> hf_weights_files = {};
+    for (auto const &pattern : allow_patterns) {
+        for (auto const& entry : fs::directory_iterator(hf_folder)) {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please move ``&`` next to ``entry``, i.e., ``for (auto const &entry``.

File: mako/utils/transformers.cc [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            if (entry.path().extension() == pattern) {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please use ``std::string::compare`` for string comparison.

File: mako/utils/transformers.cc [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                hf_weights_files.push_back(entry.path().string());
+            }
+        }
+
+        if (!hf_weights_files.empty()) {
+            if (pattern == ".safetensors") {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please use ``std::string::compare`` for string comparison.

**Review by 9rum — CHANGES_REQUESTED** [review]

File: mako/utils/transformers.cc [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mako/utils/transformers.h"
+
+#include <vector>
+#include <iostream>
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please remove unused ``#include`` directives such as ``iostream`` and ``fstream``.

File: mako/utils/transformers.cc [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+#include <fstream>
+#include <tuple>
+#include <string>
+#include <filesystem>
+#include <algorithm>
+
+namespace fs = std::filesystem;
```
Comment (9rum) [inline mako/utils/transformers.cc]: The ``namespace fs`` is already aliased in ``mako/utils/transformers.h``. This line can be removed.

**Review by 9rum — CHANGES_REQUESTED** [review]

File: mako/utils/transformers.cc (hunk 2 above)
Comment (fa-ina-tic) [reply mako/utils/transformers.cc]: First of all, thanks for your comment @9rum!

Using `static` would be great. I'll edit my code to use `static`.
But using `inline` don't look good to me.

reasons
1. We only use `prepare_hf_model_weights` once when we get model weights' file's path and it is not a simple, small function. Using inline makes code larger and might cause thrashing problem.
2. `snapshot_download` that I planned to implement have large I/O works(it download model from huggingface) and using inline function can cause blocking behavior.

It don't cause serious performance problem now, but I think using inline function can be a future-obstacles.

Here is references of my thoughts. Feel free to comment your opinion:)
- [drdobbs](https://www.drdobbs.com/to-inline-or-not-to-inline/184405660)
- [isocpp.org](https://isocpp.org/wiki/faq/inline-functions#overview-inline-fns)

Comment (9rum) [reply mako/utils/transformers.cc]: I think you may have some misunderstandings about inline functions.
Inlining is a common optimization strategy for reducing ``jmp`` and ``mov`` instructions, by copying the entire inlined code to the caller function. Thus **inlining is not related to thrashing or blocking problem**.
As mentioned in your references, such code copying may reduce a few instruction calls, while making the generated executable size larger (i.e., there is a trade-off between program's speed and size).
However, the executable size does not matter much in the modern HPC environments, so I suggest to inline the function for better speedup.

File: mako/utils/transformers.cc [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+/// \param load_format Format of the model to load. Must be one of ``"auto"``, ``"safetensors"``, ``"pt"``, or ``"npcache"``.
+/// \param fall_back_to_pt If ``true``, will always allow pt format.
+/// \param revision An optional Git revision id which can be a branch name, a tag, or a commit hash.
+/// \return Pairs of name and weight of the loaded model.
+static std::tuple<std::string, std::vector<std::string>, bool> prepare_hf_model_weights(
+        std::string model_name_or_path,
+	std::optional<std::string> cache_dir = std::nullopt,
+        std::string load_format              = "auto",
+        bool fall_back_to_pt                 = true,
+        std::optional<std::string> revision  = "None") {
```
Comment (9rum) [inline mako/utils/transformers.cc]: Please make the default value of ``revision`` to ``std::nullopt``, as same as ``cache_dir``.

File: mako/utils/transformers.cc [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    std::vector<std::string> allow_patterns;
+    try {
+        std::vector<std::string> format_allowance = {"auto", "safetensors", "pt", "npcache"};
+        if (std::find(format_allowance.begin(), format_allowance.end(), load_format) == format_allowance.end())
+            throw std::invalid_argument("invalid load_format, must be one of : auto, safetensors, pt, npcache");
+    } catch(std::invalid_argument &e) {
+        std::cerr << e.what();
+    }
+
+    if (load_format.compare("auto")) {
```
Comment (9rum) [inline mako/utils/transformers.cc]: I think this line means ``load_format`` is equal to ``"auto"``, but the actual behavior is the opposite.
Please fix to ``load_format.compare("auto") == 0`` and the below lines as same.

**Review by 9rum — CHANGES_REQUESTED** [review]

**Review by 9rum — APPROVED** [review]
LGTM

**Pull request merged**